[16.0][ADD] hr_timesheet_overtime_rate_begin_end

carmenbianca opened the pull request:
## Description

Compatibility module, depends on https://github.com/OCA/timesheet/pull/692

Built on top of #41

## Odoo task (if applicable)

T12629

## Checklist before approval

- [x] Tests are present (or not needed).
- [x] Credits/copyright have been changed correctly.
- [ ] (If a new module) Moving this to OCA has been considered.

carmenbianca committed:
[ADD] hr_timesheet_overtime_begin_end

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
I _think_ this correctly addresses the need in the internal task.

Backporting this to v12 may be a touch tricky because v12 doesn't use compute for `unit_amount`, but it's probably fine.

huguesdk commented on hr_timesheet_overtime_begin_end/models/account_analytic_line.py:
```diff
@@ -0,0 +1,28 @@
+class AnalyticLine(models.Model):
+    _inherit = "account.analytic.line"
+
+    @api.model
+    def _add_unit_amount_to_vals_list(self, vals_list):
+        # This is a kind of strange method. Like in hr_timesheet_overtime's
+        # create method, the value of unit_amount passed to create is _not_ the
+        # value written to the new record. Instead, the rate is applied on top
+        # of the provided value. This seems a little unusual/implicit/magical to
+        # me.
+        super()._add_unit_amount_to_vals_list(vals_list)
+        for vals in vals_list:
+            if vals.get("unit_amount") and (date := vals.get("date")):
+                vals["unit_amount"] *= self.rate_for_date(date)
+        return
```
if the default value is correctly handled as explained in [my comment](https://github.com/OCA/timesheet/pull/692#discussion_r1650579647) on OCA/timesheet#692, this function shouldn’t be needed anymore.

huguesdk commented on hr_timesheet_overtime_begin_end/models/account_analytic_line.py:
```diff
@@ -0,0 +1,28 @@
+
+    def unit_amount_from_start_stop(self):
+        result = super().unit_amount_from_start_stop()
+        if self.date:
```
`date` is a required field, why the test?

huguesdk commented on hr_timesheet_overtime_begin_end/tests/test_analytic_line.py:
```diff
@@ -0,0 +1,47 @@
+            "date": "2024-01-01",  # Monday
+            "time_start": 10.0,
+            "time_stop": 12.0,
+            "user_id": self.user.id,
+            "account_id": self.analytic.id,
+            "amount": -60.0,
```
same remark as [here](https://github.com/OCA/timesheet/pull/691#discussion_r1650683667).

huguesdk reviewed (changes requested):
just some changes depending on my other reviews in [OCA/timesheet](https://github.com/OCA/timesheet), otherwise lgtm.

carmenbianca commented:
Temporarily rebased on top of #37

carmenbianca commented:
Ugh this is terrible. I wrote a test that fails: `test_rate_not_double_applied`.

Problem:

The rate is applied _twice_. Once on the transient record that is shown to the user before it is actually created, and then _again_ when the transient record is written to the database.

There are two solutions that I can see:

1. Somehow add an exception in the `create` method such that the rate isn't doubly applied.
2. Fix `hr_timesheet_overtime` so that it doesn't subtly override the values you give to `write`/`create`.

carmenbianca committed:
[REF] hr_timesheet_overtime: Add hours_worked field

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] hr_timesheet_overtime_rate: Factored out of hr_timesheet_overtime

This functionality was completely unlinked from the rest of
hr_timesheet_overtime.

I would have done this in two commits, but then the files wouldn't move
over cleanly, losing history.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REM] hr_timesheet_overtime_rate: Make summary view read-only

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] hr_timesheet_overtime_rate: Populate hours_worked on installation

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] hr_timesheet_overtime_rate: Rate acquisition is no longer an @api.model method

This makes it easier to extend with additional available information on
the record.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk commented on hr_timesheet_overtime_rate_begin_end/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1 @@
+Make the two modules compatible.
```
i know it’s obvious if we look into the manifest, but stating the module names here would be useful for readers, and also in the manifest, as the description is shown in odoo without listing the dependencies.

huguesdk reviewed (changes requested):
in the timesheet lines tree, the `hours_worked` field appears before `time_start` and `time_stop`, while it would be more logical if it would appear after them, between `time_stop` and `unit_amount`.

carmenbianca committed:
[IMP] hr_timesheet_overtime_rate: Improve code aesthetics

No functional differences; code is a little prettier.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] hr_timesheet_overtime_rate: Decrease priority of views

By decreasing the priority, we make sure that hours_worked is as close
to unit_amount as possible.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
Fixed your field order remark in https://github.com/coopiteasy/cie-timesheet/pull/41/commits/2136f929d0f1d65f1abcf7ee3b88b67a8563aa24, different PR #41

carmenbianca committed:
[IMP] hr_timesheet_overtime_rate: Add hours_worked to task view

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk commented on hr_timesheet_overtime_rate_begin_end/models/account_analytic_line.py:
```diff
@@ -0,0 +1,37 @@
+
+    def unit_amount_from_start_stop(self):
+        result = super().unit_amount_from_start_stop()
+        result *= self.rate_for_date(self.date)
```
just tried it: in the interface it works ok, but as soon as i want to save:

```
AttributeError: 'account.analytic.line' object has no attribute 'rate_for_date'
```

replacing it with `self.get_overtime_rate()` seems to work.

carmenbianca replied:
good find!

huguesdk reviewed (changes requested):
just 1 small thing to fix, otherwise lgtm.

huguesdk reviewed (commented):
oops, approved too fast? some tests are still failing. is this because this still depends on other changes?
